File: code/scenes/src/main/java/org/fbme/scenes/controllers/ExtendedLayoutModel.java
package org.fbme.scenes.controllers;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.Function;

public class ExtendedLayoutModel<ResT, ExtT extends ResT, OrigT extends ResT> extends ROLayoutModelBase<ResT> {

    private final LayoutModel<OrigT> myBaseModel;
    private final Function<OrigT, Set<ExtT>> myExtensionsProvider;
    private final BiFunction<ExtT, Point, Point> myExtensionPositionProvider;
    private final BiFunction<ExtT, Point, Rectangle> myExtensionBoundsProvider;

    private final Map<OrigT, Set<ExtT>> myExtensions = new HashMap<>();
    private final Map<ExtT, OrigT> myExtensionOrigin = new HashMap<>();

    public ExtendedLayoutModel(
            LayoutModel<OrigT> baseModel,
            BiFunction<ExtT, Point, Point> extensionPositionProvider,
            BiFunction<ExtT, Point, Rectangle> extensionBoundsProvider,
            Function<OrigT, Set<ExtT>> extensionsProvider
    ) {
        myBaseModel = baseModel;
        myExtensionsProvider = extensionsProvider;
        myExtensionPositionProvider = extensionPositionProvider;
        myExtensionBoundsProvider = extensionBoundsProvider;

        for (OrigT component : myBaseModel.components()) {
            Point activePosition = myBaseModel.getActivePosition(component);
            addComponent(component, activePosition.x, activePosition.y);
        }

        myBaseModel.addListener(new ROLayoutModel.Listener<>() {
            public void onComponentAdded(@NotNull OrigT component, int x, int y) {
                fireAdded(component, x, y);
                addComponent(component, x, y);
            }

            public void onComponentRemoved(@NotNull OrigT component) {
                fireRemoved(component);
                Set<ExtT> extensions = myExtensionsProvider.apply(component);
                myExtensions.remove(component);
                for (ExtT extension : extensions) {
                    myExtensionOrigin.remove(extension);
                    fireRemoved(extension);
                }
            }

            public void onComponentsMoved(@NotNull Set<OrigT> components, int dx, int dy, boolean completed) {
                HashSet<ResT> all = new HashSet<>(components);
                for (OrigT component : components) {
                    Set<ExtT> ext = myExtensions.get(component);
                    if (ext != null) {
                        all.addAll(ext);
                    }
                }
                fireMoved(all, dx, dy, completed);
            }
        });
    }

    private void addComponent(OrigT component, int x, int y) {
        Set<ExtT> extensions = myExtensionsProvider.apply(component);
        myExtensions.put(component, extensions);
        Point componentPosition = new Point(x, y);
        for (ExtT extension : extensions) {
            myExtensionOrigin.put(extension, component);
            Point extensionPosition = myExtensionPositionProvider.apply(extension, componentPosition);
            fireAdded(extension, extensionPosition.x, extensionPosition.y);
        }
    }


    @Override
    public Set<ResT> components() {
        HashSet<ResT> res = new HashSet<>();
        res.addAll(myBaseModel.components());
        res.addAll(myExtensionOrigin.keySet());
        return res;
    }

    @Nullable
    @Override
    public ResT findAt(int x, int y) {
        // FIXME has to employ myExtensionBoundsProvider to handle extensions also
        return myBaseModel.findAt(x, y);
    }

    @NotNull
    @Override
    public Point getActivePosition(@NotNull ResT ext) {
        if (myExtensions.containsKey(ext)) {
            return myBaseModel.getActivePosition((OrigT) ext);
        }
        Point origPosition = myBaseModel.getActivePosition(myExtensionOrigin.get(ext));
        return myExtensionPositionProvider.apply((ExtT) ext, origPosition);
    }

    @Nullable
    @Override
    public Point getTracePosition(@NotNull ResT ext) {
        if (myExtensions.containsKey(ext)) {
            return myBaseModel.getTracePosition((OrigT) ext);
        }
        Point origPosition = myBaseModel.getTracePosition(myExtensionOrigin.get(ext));
        return myExtensionPositionProvider.apply((ExtT) ext, origPosition);
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/ExtendedLayoutModel.kt
package org.fbme.scenes.controllers

import java.awt.Point
import java.util.function.BiFunction
import java.util.function.Function

class ExtendedLayoutModel<ResT, ExtT : ResT, OrigT : ResT>(
    private val baseModel: LayoutModel<OrigT>,
    private val extensionPositionProvider: BiFunction<ExtT, Point, Point>,
    private val extensionsProvider: Function<OrigT, Set<ExtT>>
) : ROLayoutModelBase<ResT>() {
    private val extensions: MutableMap<OrigT, Set<ExtT>> = HashMap()
    private val extensionOrigin: MutableMap<ExtT, OrigT> = HashMap()
    private fun addComponent(component: OrigT, x: Int, y: Int) {
        val extensions = extensionsProvider.apply(component)
        this.extensions[component] = extensions
        val componentPosition = Point(x, y)
        for (extension in extensions) {
            extensionOrigin[extension] = component
            val extensionPosition = extensionPositionProvider.apply(extension, componentPosition)
            fireAdded(extension, extensionPosition.x, extensionPosition.y)
        }
    }

    override fun components(): Set<ResT> {
        val res = HashSet<ResT>()
        res.addAll(baseModel.components())
        res.addAll(extensionOrigin.keys)
        return res
    }

    override fun findAt(x: Int, y: Int): ResT? {
        // FIXME has to employ myExtensionBoundsProvider to handle extensions also
        return baseModel.findAt(x, y)
    }

    override fun getActivePosition(component: ResT): Point {
        if (extensions.containsKey(component)) {
            return baseModel.getActivePosition(component as OrigT)
        }
        val origPosition = baseModel.getActivePosition(extensionOrigin[component] ?: error("Component not found"))
        return extensionPositionProvider.apply(component as ExtT, origPosition)
    }

    override fun getTracePosition(component: ResT): Point? {
        if (extensions.containsKey(component)) {
            return baseModel.getTracePosition(component as OrigT)
        }
        val origPosition = baseModel.getTracePosition(extensionOrigin[component] ?: error("Component not found"))
        return origPosition?.let { extensionPositionProvider.apply(component as ExtT, origPosition) }
    }

    init {
        for (component in baseModel.components()) {
            val activePosition = baseModel.getActivePosition(component)
            addComponent(component, activePosition.x, activePosition.y)
        }
        baseModel.addListener(object : ROLayoutModel.Listener<OrigT> {
            override fun onComponentAdded(component: OrigT, x: Int, y: Int) {
                fireAdded(component, x, y)
                addComponent(component, x, y)
            }

            override fun onComponentRemoved(component: OrigT) {
                fireRemoved(component)
                val extensions = extensionsProvider.apply(component)
                this@ExtendedLayoutModel.extensions.remove(component)
                for (extension in extensions) {
                    extensionOrigin.remove(extension)
                    fireRemoved(extension)
                }
            }

            override fun onComponentsMoved(components: Set<OrigT>, dx: Int, dy: Int, completed: Boolean) {
                val all = HashSet<ResT>(components)
                for (component in components) {
                    val ext = extensions[component]
                    if (ext != null) {
                        all.addAll(ext)
                    }
                }
                fireMoved(all, dx, dy, completed)
            }
        })
    }
}
